### PR TITLE
Guitaskqueue taskmanager

### PR DESCRIPTION
**Pull request by whirm**
Drop TimedTasQueue and GUITaskQueue and use TwistedRawServer instead.

### CODE REVIEW

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by NielsZeilemaker** [comment]
@whirm I was looking at the changes, and similar to the other discussion we had, I'm would remove most of the blocking_ decorators. Those are really dangerous.

File: Tribler/Core/Utilities/configparser.py [hunk 1]
```diff
@@ -15,59 +15,58 @@ class CallbackConfigParser(RawConfigParser):
     def __init__(self, *args, **kwargs):
         RawConfigParser.__init__(self, *args, **kwargs)
         self.callback = None
-        self.lock = RLock()
 
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Core/Utilities/configparser.py]: Removing the lock and replacing it with a blocking call is worse than before. I would revert these changes.

File: Tribler/Core/Session.py [hunk 2]
```diff
@@ -296,25 +305,42 @@ def set_download_states_callback(self, usercallback, getpeerlist=None):
     #
     # Config parameters that only exist at runtime
     #
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: This should be blocking

File: Tribler/Core/Session.py [hunk 3]
```diff
@@ -174,15 +182,16 @@ def get_instance(*args, **kw):
...
-    get_instance = staticmethod(get_instance)
 
+    @staticmethod
+    @blocking_call_on_reactor_thread
     def has_instance():
         return Session.__single is not None
-    has_instance = staticmethod(has_instance)
 
+    @staticmethod
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: No need to make this blocking, if you want you could have a singleton_lock

File: Tribler/Core/Session.py [hunk 4]
```diff
@@ -174,15 +182,16 @@ def get_instance(*args, **kw):
         if Session.__single is None:
             Session(*args, **kw)
         return Session.__single
-    get_instance = staticmethod(get_instance)
 
+    @staticmethod
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: No need for blocking

File: Tribler/Core/Session.py [hunk 5]
```diff
@@ -166,6 +172,8 @@ def prestart(self):
     #
     # Class methods
     #
+    @staticmethod
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: No need for blocking

File: Tribler/Core/Session.py [hunk 6]
```diff
@@ -58,7 +63,7 @@ def __init__(self, scfg=None, ignore_singleton=False, autoload_discovery=True):
             if Session.__single:
                 raise RuntimeError("Session is singleton")
             Session.__single = self
-
+        assert isInIOThread()
```
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: Why does this need to be constructed on the ioThread again?

File: Tribler/Core/SessionConfig.py [hunk 7]
```diff
@@ -161,10 +167,16 @@ def set_permid_keypair_filename(self, keypairfilename):
         """
         self.sessconfig.set(u'general', u'eckeypairfilename', keypairfilename)
 
+    @blocking_call_on_reactor_thread
     def get_permid_keypair_filename(self):
         """ Returns the filename of the Session's keypair.
         @return An absolute path name. """
-        return self.sessconfig.get(u'general', u'eckeypairfilename')
+        file_name = self.sessconfig.get(u'general', u'eckeypairfilename')
+        if not file_name:
```
Comment (NielsZeilemaker) [inline Tribler/Core/SessionConfig.py]: What happened here?

File: Tribler/Core/SessionConfig.py [hunk 8]
```diff
@@ -618,12 +630,12 @@ def get_default_config_filename(state_dir):
         return os.path.join(state_dir, STATEDIR_SESSCONFIG)
 
 
-
 class SessionStartupConfig(SessionConfigInterface):
 
     """ Class to configure a Session """
 
     def __init__(self, sessconfig=None):
+        assert isInIOThread()
```
Comment (NielsZeilemaker) [inline Tribler/Core/SessionConfig.py]: Why does this need to be on the ioThread?

File: Tribler/Main/Dialogs/GUITaskQueue.py [hunk 9]
```diff
@@ -21,11 +21,11 @@ def __init__(self):
...
         TimedTaskQueue.__init__(self, nameprefix="GUITaskQueue")
 
+    @staticmethod
     def getInstance(*args, **kw):
         if GUITaskQueue.__single is None:
             GUITaskQueue(*args, **kw)
         return GUITaskQueue.__single
-    getInstance = staticmethod(getInstance)
 
     def delInstance(*args, **kw):
```
Comment (NielsZeilemaker) [inline Tribler/Main/Dialogs/GUITaskQueue.py]: Then update the delInstance as well to a decorated version

File: Tribler/Test/API/test_seeding.py [hunk 10]
```diff
@@ -32,6 +38,7 @@ def setUp(self):
         self.seeding_event = threading.Event()
         self.downloading_event = threading.Event()
 
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Test/API/test_seeding.py]: Why does this need to be on the reactor thread?

File: Tribler/Test/test_as_server.py [hunk 11]
```diff
@@ -152,7 +155,10 @@ def setUp(self):
 
         self.quitting = False
 
-        self.session = Session(self.config)
+        @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Test/test_as_server.py]: I guess this is due to the assert isIOThread stuff, doesn't make a lot of sense

File: Tribler/Test/test_as_server.py [hunk 12]
```diff
@@ -166,8 +172,10 @@ def setUp(self):
 
         self.annotate(self._testMethodName, start=True)
 
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Test/test_as_server.py]: Same here, why must the preSession be on the ioThread?

File: Tribler/Core/SessionConfig.py (hunk 7 above)
Comment (whirm) [inline Tribler/Core/SessionConfig.py]: I did that a while ago, I don't remember anymore, I guess it was getting ```None```. I will try removing that and seeing if I can reproduce whatever issue I had.

**Comment by whirm** [comment]
@NielsZeilemaker about the whole pushing stuff to the reactor thread: I was seeing if I could move as much logic to the reactor thread as possible easily. It won't be that easy so I'm leaving it for a future PR/

**Comment by NielsZeilemaker** [comment]
@whirm that makes a lot of sense. However, I would go about it the other way. By removing all other threads, this eventually will result in the same endpoint.

**Comment by NielsZeilemaker** [comment]
what about removing the taskqueue and the guitaskqueue and switching to the twistedrawserver.add_task_in_thread

File: Tribler/Core/Session.py [hunk 13]
```diff
@@ -106,21 +110,22 @@ def set_and_create_dir(dirname, setter, default_dir):
             #
             # 1. keypair
             #
-            pairfilename = os.path.join(scfg.get_state_dir(), 'ec.pem')
```
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: Ah, I see you pushed this logic into sessionconfig

Comment (whirm) [inline Tribler/Core/Session.py]: Ah, there it is! :D

**Comment by whirm** [comment]
> what about removing the taskqueue and the guitaskqueue and switching to the twistedrawserver.add_task_in_thread

Yes, I was thinking about having a few utility methods to deal with this on TaskManager but maybe we could just keep it on the rawserver for the time being.

BTW, why is the rawserver called rawserver?

**Comment by whirm** [comment]
@NielsZeilemaker I'll this pass and then proceed with the refactoring to use the rawserver.

File: Tribler/Core/Session.py (hunk 2 above)
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: This should not be blocking :smile:

File: Tribler/Core/Session.py [hunk 14]
```diff
@@ -296,25 +301,36 @@ def set_download_states_callback(self, usercallback, getpeerlist=None):
...
+          startup and due to the Dialback mechanism, this method may return
+          False while the Session is actually already reachable. Note that True
+          doesn't mean the Session is reachable from the open Internet, could just
+          be from the local (otherwise firewalled) LAN.
+          @return A boolean. """
+
+        # Arno, LICHT: make it throw exception when used in LITE versie.
+        raise NotYetImplementedException()
+
+    @blocking_call_on_reactor_thread
```
Comment (NielsZeilemaker) [inline Tribler/Core/Session.py]: Same here

**Comment by NielsZeilemaker** [comment]
It's getting there, my current worries are now that the TimedTaskQueue used to be a separate thread, and now it's running on the reactor. It might be better to run the tasks on the reactor thread pool instead.

**Comment by whirm** [comment]
@NielsZeilemaker yes, now that is passing, I guess we can directly migrate to the rawserver.add_task_in_thread()

**Comment by NielsZeilemaker** [comment]
Yes, that would be a nice improvement. And maybe lateron rename rawserver to something else.

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
Ok, I'll squash some commits and rename the rawserver too. What about PooledTaskManager or ThreadPoolTaskManager?

**Comment by whirm** [comment]
Rebase and cleanup done, only the class renaming is left.

**Comment by whirm** [comment]
Github is being silly and sorts the commits by creation date, but https://github.com/Tribler/tribler/commit/6d4fe018b6c09eb8a89a0af50061ce80cefe31ca is the last one really.

File: Tribler/Main/tribler_main.py [hunk 15]
```diff
@@ -216,7 +211,7 @@ def __init__(self, params, installdir, autoload_discovery=True,
 
             # Schedule task for checkpointing Session, to avoid hash checks after
             # crashes.
-            self.guiserver.add_task(self.guiservthread_checkpoint_timer, SESSION_CHECKPOINT_INTERVAL)
+            session.lm.rawserver.call_in_thread(SESSION_CHECKPOINT_INTERVAL, self.guiservthread_checkpoint_timer)
```
Comment (NielsZeilemaker) [inline Tribler/Main/tribler_main.py]: In the gui I would use the start_worker stuff and then specify that it should run in the threadpool.

**Comment by whirm** [comment]
Ready if you guys like the new name :)

**Comment by LipuFei** [comment]
@whirm @NielsZeilemaker Looks good :)

**Comment by whirm** [comment]
@LipuFei rebased and ready when it passes

**Comment by whirm** [comment]
retest this please

**Comment by LipuFei** [comment]
@whirm I will merge this one. That PyDeadObject problem in ListBody happens quite often these days, but I have no idea where it comes from. The backtrace is not informative enough.

**Comment by whirm** [comment]
@LipuFei yes, maybe we should have another shot at separating the GUI into a component that can be started and stopped by itself. I have a branch with some work on that but It'll need heavy refactoring. Let's talk about it on Monday and decide what to do next.
